### ISSUE TITLE
New Rule CHECK_ROOT_VOLUME_ENCRYPTION

### ISSUE THREAD

**Issue by phaughney22** [awslabs/aws-config-rules#260]
'''
#####################################
#####################################

Rule Name:
    CHECK_ROOT_VOLUME_ENCRYPTION

Description:
    Check whether the root volume associated with an instance is encrypted or not.

Trigger:
    Configuration Change on AWS::EC2::Instance

Reports on:
    AWS::EC2::Instance

Rule Parameters:
    None

Feature:
  In order to: protect the data contained on EC2 instances' root volumes
           As: a Security Officer
       I want: to verify that the root volume is encrypted on any EC2 instance in the environment.

Scenarios:

  Scenario 1:
  Given: The no instances are returned
   Then: Return NOT_APPLICABLE

  Scenario 2:
  Given: The root volume has the "Encrypted" key set as False
   Then: Return NON_COMPLIANT on this EC2 Instance

  Scenario 3:
  Given: The root volume has the "Encrypted" key set as False
    And: Additional volume attachments have the "Encrypted" key set to True
   Then: Return NON_COMPLIANT on this EC2 Instance

  Scenario 4:
  Given: The root volume has the "Encrypted" key set as False
    And: Additional volume attachments have the "Encrypted" key set to False
   Then: Return NON_COMPLIANT on this EC2 Instance

  Scenario 5:
  Given: The root volume has the "Encrypted" key set as True
   Then: Return COMPLIANT on this EC2 Instance

  Scenario 6:
  Given: The root volume has the "Encrypted" key set as True
    And: Additional volume attachments have the "Encrypted" key set to True
   Then: Return COMPLIANT on this EC2 Instance

  Scenario 7:
  Given: The root volume has the "Encrypted" key set as True
    And: Additional volume attachments have the "Encrypted" key set to False
   Then: Return COMPLIANT on this EC2 Instance

'''

**Comment by phaughney22** [comment]
Please label as "feedback requested"

**Comment by jongogogo** [comment]
Hi. What is the difference with https://docs.aws.amazon.com/config/latest/developerguide/encrypted-volumes.html ?

**Comment by phaughney22** [comment]
The one you linked checks all attached volumes. This config rule checks specifically if the volume that is attached as a root volume is encrypted or not. In this rule you can have multiple attachments but the only one that matters for compliance is the root volume.

**Comment by jongogogo** [comment]
Understood, thanks. 

I am not understanding the use case where this would be appropriate. Is it cost consideration to check only for root volume? Naturally, I'd recommend to encrypt all volumes. In addition, root volume tend to be OS only, and secondary volumes tend to be data.

**Comment by phaughney22** [comment]
The thought process behind this rule stemmed from government compliance rules where a control requires that a root volume of an instance is encrypted. A main difference between the linked AWS backed rule and this rule as that the AWS backed rule returns volumes as the compliance resource. This rule returns an instance as compliant or not. While the encrypted-volumes and this rule ultimately check the same metadata on the volume, this rule will be beneficial in instances where there are a significant amount of volumes attached to instances. A particular instance, instead of volume, can be quickly identified as either compliant or not and further remediation can be taken on the instance level such as shutting the instance down.

**Comment by jongogogo** [comment]
I remember we had a similar PR #174 . It check all volume though, not only the root volume.

**Comment by phaughney22** [comment]
Correct, that PR is closer to this one in that it is instance invoked rather than volume invoked and does not account for which volumes need to be compliant. If you believe that this rule is too specific, you can certainly close it out.
